backend/poll.go

```diff
@@ -0,0 +1,184 @@
+// Copyright Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package backend
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"golang.org/x/net/context"
+	"google.golang.org/appengine"
+	"google.golang.org/appengine/datastore"
+	"net/http"
+)
+
+const (
+	ERROR_CASE_POLL  = "error"
+	POLL_SAMPLE_PATH = "/" + CATEGORY_SAMPLE_TEMPLATES + "/poll/"
+)
+
+type PollForm struct {
+	ClientId string
+	Answer   string
+}
+
+type Result struct {
+	Percentage []int
+	Answer     string
+}
+
+type Answer struct {
+	AnswerText string
+	Votes      int
+}
+
+func InitPollSample() {
+	http.HandleFunc(POLL_SAMPLE_PATH+"submit-xhr", submitXHR)
+}
+
+func getAnswersFromDatastore(context context.Context) ([]Answer, error) {
+	query := datastore.NewQuery("Answer")
+	answers := make([]Answer, 0, 4)
+	if _, err := query.GetAll(context, &answers); err != nil {
+		return answers, err
+	}
+	return answers, nil
+}
+
+func saveAnswerIntoDatastore(context context.Context, answerText string, voteCount int) error {
+	answer := &Answer{
+		AnswerText: answerText,
+		Votes:      voteCount,
+	}
+	answerKey := datastore.NewKey(context, "Answer", answerText, 0, nil)
+	_, error := datastore.Put(context, answerKey, answer)
+	if error != nil {
+		return error
+	}
+	return nil
+}
+
+func updateAnswerCount(answers []Answer, answerText string) int {
+	voteCount := 1
+	found := false
+	for i := 0; i < len(answers); i++ {
+		if answers[i].AnswerText == answerText {
+			voteCount = answers[i].Votes + 1
+			answers[i].Votes = voteCount
+			found = true
+			break
+		}
+	}
+	if !found {
+		answers = append(answers, Answer{
+			AnswerText: answerText,
+			Votes:      voteCount,
+		})
+	}
+	return voteCount
+}
+
+func calculatePollResults(w http.ResponseWriter, ctx context.Context, pollForm PollForm) ([]Result, error) {
+	var results []Result
+	err := datastore.RunInTransaction(ctx, func(ctx context.Context) error {
+		//create keys
+		clientIdKey := datastore.NewKey(ctx, "PollForm", pollForm.ClientId, 0, nil)
+		answerKey := datastore.NewKey(ctx, "Answer", pollForm.Answer, 0, nil)
```

Comment:
I think this is what IncompleteKeys are for - if you just want to query for an object key. Also replace for other queries.